tools/audio/extraction/audiobank_file.py

```diff
@@ -0,0 +1,963 @@
+        else:
+            return f"INST_{program_number}"
+
+    def drum_grp_name(self, index):
+        if self.extraction_drums_info is not None:
+            return self.extraction_drums_info[index]
+        else:
+            return f"DRUM_{index}"
+
+    def effect_name(self, index):
+        if self.extraction_effects_info is not None:
+            return self.extraction_effects_info[index]
+        else:
+            return f"EFFECT_{index}"
+
+    def envelopes_to_xml(self, xml : XMLWriter):
+        if len(self.envelopes) == 0:
+            return
+
+        xml.write_start_tag("Envelopes")
+
+        for i,(offset,env) in self.sorted_envelopes():
+            env : Envelope
+            env.to_xml(xml, self.envelope_name(i))
+
+        xml.write_end_tag()
+
+    def samples_to_xml(self, xml : XMLWriter):
+        if len(self.sample_headers) == 0:
+            return
+
+        xml.write_start_tag("Samples")
+
+        # Emit these in the order the sample headers appear in the soundfont
+        for i,(offset,sample_header) in self.sorted_sample_headers():
+            sample_header : SoundFontSample
+            sample_header.to_xml(xml, self.lookup_sample_name(sample_header))
+
+        xml.write_end_tag()
+
+    def sfx_to_xml(self, xml : XMLWriter):
+        if len(self.sfx) == 0:
+            return
+
+        xml.write_start_tag("Effects")
+
+        for i,sfx in enumerate(self.sfx):
+            sfx.to_xml(xml, self.effect_name(i), self.sample_name_func)
+
+        xml.write_end_tag()
+
+    def drums_to_xml(self, xml : XMLWriter):
+        if len(self.drums) == 0:
+            return
+
+        xml.write_start_tag("Drums")
+
+        for i,drum_grp in enumerate(self.drum_groups):
+            if isinstance(drum_grp, list):
+                for _ in range(len(drum_grp)):
+                    xml.write_element("Drum")
+            else:
+                drum_grp : DrumGroup
+                drum_grp.to_xml(xml, self.drum_grp_name(i), self.sample_name_func, self.envelope_name_func)
+
+        xml.write_end_tag()
+
+    def instruments_to_xml(self, xml : XMLWriter):
+        if len(self.instruments) == 0:
+            return
+
+        xml.write_start_tag("Instruments")
+
+        # Write in struct order
+        for instr in sorted(self.instruments, key=lambda instr : instr.struct_index):
+            instr : Instrument
+            name = self.instrument_name(instr.program_number) if not instr.unused else None
+            instr.to_xml(xml, name, self.sample_name_func, self.envelope_name_func)
+
+        xml.write_end_tag()
+
+    def to_xml(self, name, samplebanks_base):
+        xml = XMLWriter()
+
+        start = {
+            "Name"        : name,
+            "Index"       : self.bank_num,
+            "Medium"      : self.table_entry.medium.name,
+            "CachePolicy" : self.table_entry.cache_policy.name,
+            "SampleBank"  : f"$(BUILD_DIR)/{samplebanks_base}/{self.bank1.file_name}.xml",
```

Comment:
I'm a bit saddened that the `sed` hack is proliferating. What about having the audio assembler tools take some kind of "base directory" or "include path" instead? I think it would be fine for modding if all audio assets were copied into the build dir, but maybe I'm missing something. And I understand if you don't think it's worth changing at this point

Reply:
I remember thinking about amending the tools to have a base path, but I ended up not wanting to increase the complexity of the C tools with a bunch of path manipulations. I also wrote the C tools before the build directories changed, so the code was already (essentially) done at that point.